File: VB/PivotGridFormatCondition/SimpleConditionDesigner.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports DevExpress.XtraGrid
Imports DevExpress.XtraPivotGrid

Namespace FormatCondititonsDesigner
	Partial Public Class SimpleConditionDesigner
		Inherits DevExpress.XtraEditors.XtraForm
		Private privateSourcePivot As PivotGridControl
		Protected Friend Property SourcePivot() As PivotGridControl
			Get
				Return privateSourcePivot
			End Get
			Set(ByVal value As PivotGridControl)
				privateSourcePivot = value
			End Set
		End Property
		Private privateSourceField As PivotGridField
		Protected Friend Property SourceField() As PivotGridField
			Get
				Return privateSourceField
			End Get
			Set(ByVal value As PivotGridField)
				privateSourceField = value
			End Set
		End Property


		Public Sub New()
			InitializeComponent()
			AddHandler listBoxConditions.SelectedIndexChanged, AddressOf listBoxConditions_SelectedIndexChanged
		End Sub

		Private Sub listBoxConditions_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs)
			Dim selectedItem As ItemCondition = TryCast(listBoxConditions.SelectedItem, ItemCondition)
			condititonItemControlMain.Init(selectedItem)
			simpleButtonRemoveCondition.Enabled = listBoxConditions.Items.Count > 0
		End Sub

		Private Sub SimpleConditionDesigner_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
			Text = String.Format("Simple condition for field: [{0}]", SourceField.Name)
			LoadCondititons()
			simpleButtonRemoveCondition.Enabled = listBoxConditions.Items.Count > 0
		End Sub

		Private Sub LoadCondititons()
			listBoxConditions.Items.Clear()
			condititonItemControlMain.Init(Nothing)
			For Each item As PivotGridStyleFormatCondition In SourcePivot.FormatConditions
				If item.Field.Name = SourceField.Name AndAlso Not item.Condition = FormatConditionEnum.Expression Then
					AddItemToListBox(item)
				End If
			Next item
		End Sub

		Private Sub simpleButtonAddCondititon_Click(ByVal sender As Object, ByVal e As EventArgs) Handles simpleButtonAddCondititon.Click
			If SourcePivot Is Nothing Then
				Return
			End If
			listBoxConditions.Items.BeginUpdate()
			CreateNewFormatCondition()
			listBoxConditions.Items.EndUpdate()
		End Sub

		Private Sub CreateNewFormatCondition()
			Dim item As New PivotGridStyleFormatCondition()
			item.Field = SourceField
			SourcePivot.FormatConditions.Add(item)
			AddItemToListBox(item)
		End Sub

		Private Sub condititonItemControlMain_RefreshValues(ByVal sender As Object, ByVal e As EventArgs) Handles condititonItemControlMain.RefreshValues
			listBoxConditions.Refresh()
		End Sub

		Private Sub simpleButtonRemoveCondition_Click(ByVal sender As Object, ByVal e As EventArgs) Handles simpleButtonRemoveCondition.Click
			If SourcePivot Is Nothing Then
				Return
			End If
			Dim condition As ItemCondition = TryCast(listBoxConditions.SelectedItem, ItemCondition)
			listBoxConditions.Items.Remove(listBoxConditions.SelectedItem)
			SourcePivot.FormatConditions.Remove(condition.Condition)
		End Sub

		Protected Sub AddItemToListBox(ByVal newItem As PivotGridStyleFormatCondition)
			Dim newItemCondition As New ItemCondition(newItem)
			listBoxConditions.Items.Add(newItemCondition)
			listBoxConditions.SelectedItem = newItemCondition
		End Sub
	End Class
End Namespace

File: CS/PivotGridFormatCondition/SimpleConditionDesigner.cs
using System;
using System.Collections.Generic;
using DevExpress.XtraGrid;
using DevExpress.XtraPivotGrid;

namespace FormatCondititonsDesigner
{
    public partial class SimpleConditionDesigner : DevExpress.XtraEditors.XtraForm
    {
        protected internal PivotGridControl SourcePivot { get; set; }
        protected internal PivotGridField SourceField { get; set; }


        public SimpleConditionDesigner()
        {
            InitializeComponent();
            listBoxConditions.SelectedIndexChanged += new EventHandler(listBoxConditions_SelectedIndexChanged);
        }

        void listBoxConditions_SelectedIndexChanged(object sender, EventArgs e)
        {
            ItemCondition selectedItem = listBoxConditions.SelectedItem as ItemCondition;
            condititonItemControlMain.Init(selectedItem);
            simpleButtonRemoveCondition.Enabled = listBoxConditions.Items.Count > 0;
        }

        private void SimpleConditionDesigner_Load(object sender, EventArgs e)
        {
            Text = String.Format("Simple condition for field: [{0}]", SourceField.Name);
            LoadCondititons();
            simpleButtonRemoveCondition.Enabled = listBoxConditions.Items.Count > 0;
        }

        void LoadCondititons()
        {
            listBoxConditions.Items.Clear();
            condititonItemControlMain.Init(null);
            foreach (PivotGridStyleFormatCondition item in SourcePivot.FormatConditions)
            {
                if (item.Field == SourceField && item.Condition != FormatConditionEnum.Expression)
                {
                    AddItemToListBox(item);
                }
            }
        }

        private void simpleButtonAddCondititon_Click(object sender, EventArgs e)
        {
            if (SourcePivot == null) return;
            listBoxConditions.Items.BeginUpdate();
            CreateNewFormatCondition();
            listBoxConditions.Items.EndUpdate();
        }

        private void CreateNewFormatCondition()
        {
            PivotGridStyleFormatCondition item = new PivotGridStyleFormatCondition();
            item.Field = SourceField;
            SourcePivot.FormatConditions.Add(item);
            AddItemToListBox(item);
        }

        private void condititonItemControlMain_RefreshValues(object sender, EventArgs e)
        {
            listBoxConditions.Refresh();
        }

        private void simpleButtonRemoveCondition_Click(object sender, EventArgs e)
        {
            if (SourcePivot == null) return;
            ItemCondition condition = listBoxConditions.SelectedItem as ItemCondition;
            listBoxConditions.Items.Remove(listBoxConditions.SelectedItem);
            SourcePivot.FormatConditions.Remove(condition.Condition);
        }

        protected void AddItemToListBox(PivotGridStyleFormatCondition newItem)
        {
            ItemCondition newItemCondition = new ItemCondition(newItem);
            listBoxConditions.Items.Add(newItemCondition);
            listBoxConditions.SelectedItem = newItemCondition;        
        }
    }
}